ethereum/spec/src/main/java/tech/pegasys/teku/spec/schemas/SchemaRegistry.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright Consensys Software Inc., 2024
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.spec.schemas;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.HashMap;
+import java.util.Map;
+import tech.pegasys.teku.spec.SpecMilestone;
+import tech.pegasys.teku.spec.config.SpecConfig;
+import tech.pegasys.teku.spec.schemas.SchemaTypes.SchemaId;
+
+public class SchemaRegistry {
+  private final Map<SchemaId<?>, SchemaProvider<?>> providers = new HashMap<>();
+  private final SpecMilestone milestone;
+  private final SchemaCache cache;
+  private final SpecConfig specConfig;
+
+  public SchemaRegistry(
+      final SpecMilestone milestone, final SpecConfig specConfig, final SchemaCache cache) {
+    this.milestone = milestone;
+    this.specConfig = specConfig;
+    this.cache = cache;
+  }
+
+  public void registerProvider(final SchemaProvider<?> provider) {
+    providers.put(provider.getSchemaId(), provider);
+  }
+
+  @VisibleForTesting
+  boolean isProviderRegistered(final SchemaProvider<?> provider) {
+    return provider.equals(providers.get(provider.getSchemaId()));
+  }
+
+  @SuppressWarnings("unchecked")
+  public <T> T get(final SchemaId<T> schemaClass) {
+    SchemaProvider<T> provider = (SchemaProvider<T>) providers.get(schemaClass);
+    if (provider == null) {
+      throw new IllegalArgumentException(
+          "No provider registered for schema "
+              + schemaClass
+              + " or it does not support milestone "
+              + milestone);
+    }
+    T schema = cache.get(milestone, schemaClass);
+    if (schema != null) {
+      return schema;
+    }
+    final SpecMilestone effectiveMilestone = provider.getEffectiveMilestone(milestone);
+    if (effectiveMilestone != milestone) {
+      schema = cache.get(effectiveMilestone, schemaClass);
+      if (schema != null) {
+        cache.put(milestone, schemaClass, schema);
+        return schema;
+      }
+    }
+    schema = provider.getSchema(this);
+    cache.put(effectiveMilestone, schemaClass, schema);
+    if (effectiveMilestone != milestone) {
+      cache.put(milestone, schemaClass, schema);
+    }
+    return schema;
+  }
```

Comment:
We are relying on the fact that we are calling `SchemaRegistryBuilder` from a "concurrency free" part of the code. I understand this is by design but maybe we can remove that assumption (so it does not bite us in the future).

Once the registry is primed, we should always hit the cache. However, it isn't thread-safe before priming. 

What about separating the logic on `get()` to only read the cache. If we attempt to call `get()` w/o priming the registry first, we fail. We can move the logic that populates the cache to `primeRegistry()`, and make it synchronized so we know it is thread-safe no matter where it is being called from.

Reply:
Yes I agree! Generally, a better separation between priming and lookup reflects the usage we expect.

Reply:
Relooking at this, the separation can't be done.
The way this works is that, when the schema is being created, it can lookup in the registry itself so it may trigger a creation of a schema (dependency). This help us not dealing with ordering (it works as soon as there are no dependency loops).

The real point here is if we want or not a dependency loop detection. To me it can only happen if as a developer you pick the wrong schemaId (and it is compatible with the type you need in that moment). Let me think